orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/tasks/pipeline/GetPipelinesFromArtifactTask.java

```diff
@@ -0,0 +1,136 @@
+    @JsonProperty("pipelinesArtifactId") private String id;
+    @JsonProperty("pipelinesArtifact") private Artifact inline;
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public TaskResult execute(Stage stage) {
+    final PipelinesArtifactData pipelinesArtifact = stage.mapTo(PipelinesArtifactData.class);
+    Artifact resolvedArtifact = artifactResolver
+      .getBoundArtifactForStage(stage, pipelinesArtifact.getId(), pipelinesArtifact.getInline());
+    if (resolvedArtifact == null) {
+      throw new IllegalArgumentException("No artifact could be bound to '" + pipelinesArtifact.getId() + "'");
+    }
+    log.info("Using {} as the pipelines to be saved", pipelinesArtifact);
+
+    String pipelinesText = getPipelinesArtifactContent(resolvedArtifact);
+
+    Map<String, List<Map>> pipelinesFromArtifact = null;
+    try {
+      pipelinesFromArtifact = objectMapper.readValue(pipelinesText, new TypeReference<Map<String, List<Map>>>() {});
+    } catch (IOException e) {
+      log.warn("Failure parsing pipelines from {}", pipelinesArtifact, e);
+      throw new IllegalStateException(e); // forces a retry
+    }
+    final Map<String, List<Map>> finalPipelinesFromArtifact = pipelinesFromArtifact;
+    final Set<String> appNames = pipelinesFromArtifact.keySet();
+    final List newAndUpdatedPipelines = appNames.stream().flatMap(appName -> {
+      final List<Map<String, Object>> existingAppPipelines = front50Service.getPipelines(appName);
+      final List<Map> specifiedAppPipelines = finalPipelinesFromArtifact.get(appName);
+      return specifiedAppPipelines.stream().map(p -> {
+        final Map<String, Object> pipeline = p;
+        pipeline.put("application", appName);
+        final Optional<Map<String, Object>> matchedExistingPipeline = existingAppPipelines
+          .stream().filter(existingPipeline -> existingPipeline.get("name").equals(pipeline.get("name"))).findFirst();
+        matchedExistingPipeline.ifPresent(matchedPipeline -> {
+          pipeline.put("id", matchedPipeline.get("id"));
+        });
+        return pipeline;
+      }).filter(pipeline -> !pipeline.isEmpty());
+    }).collect(Collectors.toList());
+    final SavePipelinesData output = new SavePipelinesData(null, newAndUpdatedPipelines);
+    return new TaskResult(ExecutionStatus.SUCCEEDED,
+      objectMapper.convertValue(output, new TypeReference<Map<String, Object>>() {}));
+  }
+
+  private String getPipelinesArtifactContent(Artifact artifact) {
+    return retrySupport.retry(() -> {
+      Response response = oort.fetchArtifact(artifact);
+      InputStream artifactInputStream;
+      try {
+        artifactInputStream = response.getBody().in();
+      } catch (IOException e) {
+        log.warn("Failure fetching pipelines from {}", artifact, e);
+        throw new IllegalStateException(e); // forces a retry
+      }
+      try (BufferedReader rd = new BufferedReader(new InputStreamReader(artifactInputStream))) {
```

Comment:
Could do this all with `CharStreams.toString(artifactInputStream, Charsets.UTF_8)`.